[Config] Introduce find method in ArrayNodeDefinition to ease configuration tree manipulation

jschaedl opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no <!-- don't forget to update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets | #27534   <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        | tbd.

### Description

This PR introduces a new `find(string $nodePath)`method in the `ArrayNodeDefinition` class, which helps you finding the right node to prepend configuration to ease configuration tree manipulation.

### How to use it
```php
class Configuration implements ConfigurationInterface
{
    public function getConfigTreeBuilder()
    {
        ...

        $rootNode
            ->children()
                ->arrayNode('social_media_channels')
                    ->children()
                        ->booleanNode('enable')->end()
                        ->arrayNode('twitter')->end()
                        ->arrayNode('facebook')->end()
                        ->arrayNode('instagram')->end()
                    ->end()
                ->end()
            ->end()
        ;

        $this->changeSocialMediaChannelConfiguration($rootNode->find('social_media_channels.enable'));
        $this->addTwitterConfiguration($rootNode->find('social_media_channels.twitter'));
        $this->addFacebookConfiguration($rootNode->find('social_media_channels.facebook'));
        $this->addInstagramConfiguration($rootNode->find('social_media_channels.instagram'));

        return $treeBuilder;
    }

    private function changeSocialMediaChannelConfiguration(NodeDefinition $node)
    {
        $node
            ->defaultTrue()
        ;
    }

    private function addTwitterConfiguration(NodeDefinition $node)
    {
        $node
            ->children()
                ->integerNode('client_id')->end()
                ->scalarNode('client_secret')->end()
            ->end()
        ;
    }

    private function addFacebookConfiguration(NodeDefinition $node)
    {
        $node
            ->children()
                ->integerNode('client_id')->end()
                ->scalarNode('client_secret')->end()
            ->end()
        ;
    }

    private function addInstagramConfiguration(NodeDefinition $node)
    {
        $node
            ->children()
                ->integerNode('client_id')->end()
                ->scalarNode('client_secret')->end()
            ->end()
        ;
    }
}

```

jschaedl committed:
[Config] Introduce find method in ArrayNodeDefinition to ease configuration tree manipulation

OskarStark commented on src/Symfony/Component/Config/Definition/Builder/NodeFinder.php:
```diff
@@ -0,0 +1,56 @@
+ */
+class NodeFinder
+{
+    public function find(string $nodePath, NodeDefinition $rootNode)
```
Please add a description of the format of „$nodePath“

jschaedl replied:
Done

OskarStark commented on src/Symfony/Component/Config/Definition/Builder/NodeFinder.php:
```diff
@@ -0,0 +1,56 @@
+/**
+ * @author Jan Schädlich <jan.schaedlich@sensiolabs.de>
+ */
+class NodeFinder
```
Please add a class description

OskarStark commented on src/Symfony/Component/Config/Definition/Builder/NodeFinder.php:
```diff
@@ -0,0 +1,56 @@
+        $explodedNodePaths = explode('.', $nodePath);
+
+        if (!array_key_exists($explodedNodePaths[0], $rootNode->getChildNodeDefinitions())) {
+            throw new \RuntimeException(
```
The whole exception should be on one line, Symfony uses long lines CS

jschaedl replied:
Fixed

OskarStark commented on src/Symfony/Component/Config/Tests/Definition/Builder/NodeFinderTest.php:
```diff
@@ -0,0 +1,79 @@
+ */
+class NodeFinderTest extends TestCase
+{
+    public function testOne()
```
the method name should reflect what is tested

jschaedl replied:
Fixed

OskarStark reviewed (commented):
Nice improvement, I like it 👌🏻
I left some minor comments

OskarStark commented on src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php:
```diff
@@ -44,6 +44,11 @@ public function __construct(?string $name, NodeParentInterface $parent = null)
         $this->name = $name;
     }
 
+    public function getName()
```
A return type is possible, right?

jschaedl replied:
I added the type hint.

OskarStark commented on src/Symfony/Component/Config/Definition/Builder/NodeFinder.php:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @return NodeDefinition
+     */
+    public function find($nodePath, NodeDefinition $rootNode)
```
Please add return type

OskarStark commented on src/Symfony/Component/Config/Definition/Builder/NodeFinder.php:
```diff
@@ -0,0 +1,55 @@
+     * Finds a node defined by $nodePath within the given $rootNode.
+     *
+     * @param string         $nodePath The path of the node to find. e.g "doctrine.orm.mappings"
+     * @param NodeDefinition $rootNode
```
This line and the @return could be removed

jschaedl replied:
Done

OskarStark commented on src/Symfony/Component/Config/Tests/Definition/Builder/NodeFinderTest.php:
```diff
@@ -0,0 +1,114 @@
+     * @param string         $expectedName
+     * @param string         $expectedType
+     */
+    private function assertNode(NodeDefinition $actualNode, $expectedName, $expectedType)
```
Please add string type hint and remove the entire doc bloc

jschaedl replied:
Done

xabbuh commented on src/Symfony/Component/Config/Definition/Builder/NodeFinder.php:
```diff
@@ -0,0 +1,52 @@
+     *
+     * @param string $nodePath The path of the node to find. e.g "doctrine.orm.mappings"
+     */
+    public function find(string $nodePath, NodeDefinition $rootNode): NodeDefinition
```
What about making the method `static`?

xabbuh commented on src/Symfony/Component/Config/Definition/Builder/NodeFinder.php:
```diff
@@ -0,0 +1,52 @@
+            return $rootNode;
+        }
+
+        $explodedNodePaths = explode('.', $nodePath);
```
It looks like we do not need all segments, but that we can just fetch the first segment and perform all following statements based on string operations.

ro0NL replied:
you also need to account for `setPathSeparator()`

ro0NL commented on src/Symfony/Component/Config/Definition/Builder/NodeFinder.php:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @author Jan Schädlich <jan.schaedlich@sensiolabs.de>
+ */
+class NodeFinder
```
should be final IMHO

ro0NL commented on src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php:
```diff
@@ -44,6 +44,11 @@ public function __construct(?string $name, NodeParentInterface $parent = null)
         $this->name = $name;
     }
 
+    public function getName(): string
```
`?string`

ro0NL commented on src/Symfony/Component/Config/Definition/Builder/NodeFinder.php:
```diff
@@ -0,0 +1,52 @@
+     */
+    public function find(string $nodePath, NodeDefinition $rootNode): NodeDefinition
+    {
+        if (!$rootNode instanceof ArrayNodeDefinition) {
```
to me this shows it could also be solved using `ArrayNodeDefinition::find()`, which would avoid the additional getters right?

xabbuh replied:
Sounds good to me. Then we also don't need this class at all.

jschaedl commented:
@ro0NL @xabbuh 

I moved the `find(...)` method into the `ArrayNodeDefinition` class, took the `pathSeparator` into account and changed the path section extraction as suggested. I also adjusted the tests accordingly. 

WDYT?

xabbuh commented on src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php:
```diff
@@ -523,4 +523,32 @@ public function getChildNodeDefinitions()
+     *
+     * @param string $nodePath The path of the node to find. e.g "doctrine.orm.mappings"
+     *
+     * @return NodeDefinition
```
can be removed

xabbuh commented on src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php:
```diff
@@ -523,4 +523,32 @@ public function getChildNodeDefinitions()
+            return $this;
+        }
+
+        $firstPathSegment = strtok($nodePath, $this->pathSeparator);
```
`$pathSeparator` doesn't necessarily have to be a single character if I don't miss anything. Using more than one character as the separator would break here.

jschaedl replied:
Actually this shouldn't be a problem: https://3v4l.org/ET6V6

xabbuh replied:
But something like https://3v4l.org/Aucsi should not happen, right?

jschaedl replied:
Yes you are right. Using a separator like `.|` with a given node name like `mailer.config` would break the find() method. I added a test for this and removed the `strtok` method in favour of `strpos` and `substr` See: 9e7daa3

xabbuh commented on src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php:
```diff
@@ -523,4 +523,32 @@ public function getChildNodeDefinitions()
+
+        $firstPathSegment = strtok($nodePath, $this->pathSeparator);
+
+        if (!\array_key_exists($firstPathSegment, $this->getChildNodeDefinitions())) {
```
`!isset($this->children[$firstPathSegment])`?

ro0NL commented on src/Symfony/Component/Config/Tests/Definition/Builder/ArrayNodeDefinitionTest.php:
```diff
@@ -357,6 +359,85 @@ public function testCannotBeEmptyOnConcreteNode()
+    public function testFindShouldThrowExceptionIfNodeDoesNotExistInRootNode()
+    {
+        $this->expectException(\RuntimeException::class);
+        $this->expectExceptionMessage('Node with name "child" does not exist in the current node "root"!');
```
`!` should be a regular `.` IMHO

ro0NL commented on src/Symfony/Component/Config/Tests/Definition/Builder/ArrayNodeDefinitionTest.php:
```diff
@@ -357,6 +359,85 @@ public function testCannotBeEmptyOnConcreteNode()
+        $rootNode->find('child');
+    }
+
+    public function testFindShouldHandleComplexConfigurationProbably()
```
you mean `Properly`?

jschaedl replied:
Your are probably right 🙈 Will fix it. Thanks @ro0NL 👍

fabpot commented:
Thank you @jschaedl.
